File: 3DRadSpace/Engine3DRadSpace/Math/Rectangle.cpp
#include "Rectangle.hpp"
using namespace Engine3DRadSpace::Math;

constexpr Point Rectangle::TopLeft() const
{
    return Point(X, Y);
}

constexpr Point Rectangle::TopRight() const
{
    return Point(X + Width, Y);
}


constexpr Point Rectangle::BottomLeft() const
{
    return Point(X, Y + Height);
}

constexpr Point Rectangle::BottomRight() const
{
    return Point(X + Width, Y + Height);
}

constexpr Point Rectangle::Center() const
{
    return Point((X + Width) / 2, (Y + Height) / 2);
}


// ------------------------------------------------------------------------------ //


constexpr Vector2 RectangleF::TopLeft() const
{
    return Vector2(X, Y);
}

constexpr Vector2 RectangleF::TopRight() const
{
    return Vector2(X + Width, Y);
}


constexpr Vector2 RectangleF::BottomLeft() const
{
    return Vector2(X, Y + Height);
}

constexpr Vector2 RectangleF::BottomRight() const
{
    return Vector2(X + Width, Y + Height);
}

constexpr Vector2 RectangleF::Center() const
{
    return Point((X + Width) / 2.0f, (Y + Height) / 2.0f);
}

File: 3DRadSpace/3DRadSpace_CSharp/Math/Rectangle.cs
using System.Runtime.InteropServices;

namespace Engine3DRadSpace.Math
{
	[StructLayout(LayoutKind.Sequential)]
	public struct RectangleF : IEquatable<RectangleF>
	{
		public float X;
		public float Y;
		public float Width;
		public float Height;

		public RectangleF(float x, float y, float w, float h)
		{
			X = x;
			Y = y;
			Width = w;
			Height = h;
		}

		public override bool Equals(object? obj)
		{
			if (obj is RectangleF r)
			{
				return r.Equals(obj);
			}
			return false;
		}

		public bool Equals(RectangleF other)
		{
			return X == other.X &&
				Y == other.Y &&
				Width == other.Width &&
				Height == other.Height;
		}

		public override int GetHashCode()
		{
			return HashCode.Combine(X, Y, Width, Height);
		}

		public static bool operator ==(RectangleF left, RectangleF right)
		{
			return left.Equals(right);
		}
		public static bool operator !=(RectangleF left, RectangleF right)
		{
			return !left.Equals(right);
		}
	}

	[StructLayout(LayoutKind.Sequential)]
	public struct Rectangle : IEquatable<Rectangle>
	{
		public int X;
		public int Y;
		public int Width;
		public int Height;

		public Rectangle(int x, int y, int w, int h)
		{
			X = x;
			Y = y;
			Width = w;
			Height = h;
		}

		public bool Equals(Rectangle other)
		{
			return X == other.X &&
				Y == other.Y &&
				Width == other.Width &&
				Height == other.Height;
		}

		public override bool Equals(object? other)
		{
			if (other is Rectangle r)
			{
				return r.Equals(this);
			}
			return false;
		}

		public override int GetHashCode()
		{
			return HashCode.Combine(X, Y, Width, Height);
		}

		public static bool operator ==(Rectangle a, Rectangle b)
		{
			return a.Equals(b);
		}
		public static bool operator !=(Rectangle a, Rectangle b)
		{
			return !a.Equals(b);
		}
	}
}
